src/actions/clipboard.ts

```diff
@@ -125,93 +159,96 @@ export class Clipboard {
+    const oldCopyShortcut =
+      ShortcutRegistry.registry.getRegistry()[ShortcutItems.names.COPY];
+    if (!oldCopyShortcut)
+      throw new Error('No copy keyboard shortcut registered initially');
+
+    this.oldCopyCallback = oldCopyShortcut.callback;
+
     const copyShortcut: ShortcutRegistry.KeyboardShortcut = {
       name: Constants.SHORTCUT_NAMES.COPY,
-      preconditionFn: this.copyPrecondition.bind(this),
+      preconditionFn: oldCopyShortcut.preconditionFn,
       callback: this.copyCallback.bind(this),
-      keyCodes: [
-        createSerializedKey(KeyCodes.C, [KeyCodes.CTRL]),
-        createSerializedKey(KeyCodes.C, [KeyCodes.ALT]),
-        createSerializedKey(KeyCodes.C, [KeyCodes.META]),
-      ],
-      allowCollision: true,
+      // The registry gives back keycodes as an object instead of an array
+      // See https://github.com/google/blockly/issues/9008
+      keyCodes: Object.values(oldCopyShortcut.keyCodes ?? []),
+      allowCollision: false,
     };
+
+    ShortcutRegistry.registry.unregister(ShortcutItems.names.COPY);
     ShortcutRegistry.registry.register(copyShortcut);
   }
 
   /**
-   * Register the copy block action as a context menu item on blocks.
-   * This function mixes together the keyboard and context menu preconditions
-   * but only calls the keyboard callback.
+   * Register the copy block action as a context menu item.
+   * The context menu uses its own preconditionFn (that doesn't check
+   * if a gesture is in progress, because one always is in the context
+   * menu). It calls the copy callback that is shared between keyboard
+   * and context menu.
    */
   private registerCopyContextMenuAction() {
     const copyAction: ContextMenuRegistry.RegistryItem = {
-      displayText: (scope) => `Copy (${getShortActionShortcut('copy')})`,
+      displayText: (scope) =>
+        `Copy (${getShortActionShortcut(Constants.SHORTCUT_NAMES.COPY)})`,
+      // This is similar to the preconditionFn for the keyboard shortcut in core,
+      // but it doesn't have the check if a gesture is in progress.
       preconditionFn: (scope) => {
-        const ws = scope.block?.workspace;
-        if (!ws) return 'hidden';
-
-        return this.copyPrecondition(ws) ? 'enabled' : 'disabled';
+        const focused = scope.focusedNode;
+
+        if (!focused || !isCopyable(focused)) return 'hidden';
+
+        const workspace = focused.workspace;
+        if (
+          !workspace.isReadOnly() &&
+          isDeletable(focused) &&
+          focused.isDeletable() &&
+          isDraggable(focused) &&
+          focused.isMovable() &&
+          !focused.workspace.isFlyout
+        )
+          return 'enabled';
+
+        return 'disabled';
       },
-      callback: (scope) => {
-        const ws = scope.block?.workspace;
-        if (!ws) return;
-        return this.copyCallback(ws);
+      callback: (scope, menuOpenEvent) => {
+        if (!isCopyable(scope.focusedNode)) return false;
```

Comment:
Can this be moved to the precondition?

Reply:
it is also in the precondition, but it is here as well because if the thing is `ICopyable` it will definitely have a `workspace` property, so without that we'd have to cast in order to check the property.